src/Instance.ts

```diff
@@ -317,10 +324,10 @@ export class Marked {
 
       if (opt.async) {
         return Promise.resolve(opt.hooks ? opt.hooks.preprocess(src) : src)
-          .then(src => lexer(src, opt))
-          .then(tokens => opt.walkTokens ? Promise.all(this.walkTokens(tokens, opt.walkTokens)).then(() => tokens) : tokens)
-          .then(tokens => parser(tokens, opt))
-          .then(html => opt.hooks ? opt.hooks.postprocess(html) : html)
+          .then(src => lexer(src, opt as MarkedOptions))
+          .then(tokens => (opt as MarkedOptions).walkTokens ? Promise.all(this.walkTokens(tokens, (opt as MarkedOptions).walkTokens!)).then(() => tokens) : tokens)
+          .then(tokens => parser(tokens, opt as MarkedOptions))
+          .then(html => (opt as MarkedOptions).hooks ? (opt as MarkedOptions).hooks!.postprocess(html) : html)
```

Comment:
Something seems wrong here - we shouldn't need to type cast on every line.

Can we do it once above (preferably in a runtime safe way)?

Reply:
The problem is that `opt` has type `MarkedOptions | ResultCallback` due to the function overload. I tried to convert the code to Typescript (at first) without changing the resulting code. I added a new variable now to get rid of the casts.